src/umap/WorkQueue.hpp

```diff
@@ -69,7 +69,7 @@ class WorkQueue {
       pthread_mutex_lock(&m_mutex);
       ++m_idle_waiters;
 
-      while ( m_waiting_workers != m_max_waiting )
+      while ( ! ( m_queue.size() == 0 && m_waiting_workers == m_max_waiting ) )
```

Comment:
Not sure I understand the second part of the &&. The previous version of the code said != and this version says ==. Maybe a comment explaining this logic.

Reply:
OK, I see the outmost ! which makes the new condition equivalent to the previous version, and the the queue_size clause adds the extra condition.